### PR TITLE
[jit] Remove weak script

### PR DESCRIPTION
**Pull request by driazati**
* Deletes all weak script decorators / associated data structures / methods
   * In order to keep supporting the standard library in script, this enables recursive script on any function defined in `torch.nn`
   * Most changes in `torch/nn` are the result of `ag -Q "weak" torch/nn/ -l | xargs sed -i '/weak/d'`, only `rnn.py` needed manual editing to use the `@ignore` and `@export` to continue supporting the overloaded `forward` methods
* `Sequential`/`ModuleList` no longer need to be added to constants since they are compiled on demand

This should also fix #22212

Differential Revision: [D15988346](https://our.internmc.facebook.com/intern/diff/15988346/)

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -1774,34 +1728,28 @@ def __init__(self, optimize=True):
...
         # Create constant versions for the iterable modules
         return _create_constant_iterable_module(mod)
 
-    if methods is None:
-        methods = ('forward',)
+    if methods is None and hasattr(mod, 'forward'):
+        if mod.forward.__func__ == torch.nn.Module.forward:
+            # forward was not overrided
+            raise RuntimeError("No forward (TODO: delete this error)")
+        if not _jit_internal.is_ignored_fn(mod.forward):
```
Comment (zdevito) [inline torch/jit/__init__.py]: I think we should delete the `methods` argument. It is never used in the current codebase.

File: torch/nn/parallel/replicate.py [hunk 2]
```diff
@@ -25,7 +25,6 @@ def _is_jit_enabled():
 # Check if we can safely replicate the module.
 # there are three types of module:
 # 1. python modules
-# 2. weak python modules (nn.Module annotated by @weak_module)
```
Comment (zdevito) [inline torch/nn/parallel/replicate.py]: Comments are bogus here now, update?

**Review by zdevito — APPROVED** [review]
This looks great! Thanks for the progress on cleaning up WeakScriptModule.

Not for this PR but for follow up: we have a lot of free functions in __init__.py that really should be private methods of the new ScriptModule  (e.g. _make_strong_submodule). Also audit __init__.py for dead code once the WeakScriptModule class itself is removed.

File: torch/_jit_internal.py [hunk 3]
```diff
@@ -71,51 +56,41 @@ def env(key):
...
         elif hasattr(builtins, key):
             return getattr(builtins, key)
-        else:
-            return None
 
     return env
 
 
-def weak_script(fn, _frames_up=0):
+def createResolutionCallbackFromClosure(fn):
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/_jit_internal.py]: Right, to put it in the same place as `createResolutionCallback`

File: test/test_jit.py [hunk 4]
```diff
@@ -10931,6 +10927,7 @@ def foo(cond):
         with self.assertRaisesRegex(torch.jit.Error, "Exception"):
             foo(torch.tensor(0))
 
+    @unittest.skipIf(True, "Removing weak script")
```
[the remark this answers is not in the document]
Comment (driazati) [reply test/test_jit.py]: They're re-enabled in `TestRecursiveScript` in the latest commit

**Comment by zhangguanheng66** [comment]
Thanks for working on this issue. Once landed, I will revise our code according and test the jit mode.

**Pull request closed without merging**